site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,694 @@
+	//   conformance levels. Filter-specific configuration for such filters
+	//   is specified using the ExtensionRef field. `Type` should be set to
+	//   "ExtensionRef" for custom filters.
+	//
+	// Implementers are encouraged to define custom implementation types to
+	// extend the core API with implementation-specific behavior.
+	//
+	// If a reference to a custom filter type cannot be resolved, the filter
+	// MUST NOT be skipped. Instead, requests that would have been processed by
+	// that filter MUST receive a HTTP error response.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum=RequestHeaderModifier;RequestMirror;ExtensionRef
+	// <gateway:experimental:validation:Enum=RequestHeaderModifier;RequestMirror;ExtensionRef>
+	Type GRPCRouteFilterType `json:"type"`
+
+	// RequestHeaderModifier defines a schema for a filter that modifies request
+	// headers.
+	//
+	// Support: Core
+	//
+	// Support: Core
+	//
+	// +optional
+	RequestHeaderModifier *HTTPRequestHeaderFilter `json:"requestHeaderModifier,omitempty"`
+
+	// RequestMirror defines a schema for a filter that mirrors requests.
+	// Requests are sent to the specified destination, but responses from
+	// that destination are ignored.
+	//
+	// Support: Extended
+	//
+	// +optional
+	RequestMirror *HTTPRequestMirrorFilter `json:"requestMirror,omitempty"`
+
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "filter" behavior.  For example, resource "myroutefilter" in group
+	// "networking.example.net"). ExtensionRef MUST NOT be used for core and
+	// extended filters.
+	//
+	// Support: Implementation-specific
+	// Support: Implementation-specific
+	//
+	// +optional
+	ExtensionRef *LocalObjectReference `json:"extensionRef,omitempty"`
+}
+```
+
+In addition, two new `ProtocolType` options will be added:
+
+```diff
+--- a/apis/v1alpha2/gateway_types.go
++++ b/apis/v1alpha2/gateway_types.go
+@@ -165,6 +165,11 @@ type Listener struct {
+        //   protocol layers as described above. If an implementation does not
+        //   ensure that both the SNI and Host header match the Listener hostname,
+        //   it MUST clearly document that.
++       // * H2C: The Listener Hostname MUST match the :authority pseudo-header of the request.
++       // * H2: The Listener Hostname SHOULD match at both the TLS and HTTP
++       //   protocol layers as described above. If an implementation does not
++       //   ensure that both the SNI and :authority pseudo-header match the Listener hostname,
++       //   it MUST clearly document that.
+        //
+        // For HTTPRoute and TLSRoute resources, there is an interaction with the
+        // `spec.hostnames` array. When both listener and route specify hostnames,
+@@ -279,6 +284,12 @@ const (
+ 
+        // Accepts UDP packets.
+        UDPProtocolType ProtocolType = "UDP"
++
++       // Accepts HTTP/2 sessions over TLS.
++       H2ProtocolType ProtocolType = "H2"
```

Comment:
Because `HTTPS` doesn't necessarily imply support for HTTP/2. You could have an `HTTPS` implementation that only supports HTTP/1.1. An implementation supporting `H2` would guarantee that HTTP/2 over TLS with no upgrade from HTTP/1 works.

Reply:
It seems a bit inconsistent... so we now have:

* `HTTP`: h1, optionally h2c
* `HTTPS`: h1 or h2 (over TLS)
* `H2C`: only h2c
* `HTTP2`: only h2 (over TLS)

I can't imagine any user could see those 4 options and understand what they mean without a close inspection of the docs

Also...
> You could have an HTTPS implementation that only supports HTTP/1.1

Doesn't the spec not allow that?

```
	// Accepts HTTP/1.1 or HTTP/2 sessions over TLS.
	HTTPSProtocolType ProtocolType = "HTTPS"
```

> An implementation supporting H2 would guarantee that HTTP/2 over TLS with no upgrade from HTTP/1 works.

It guarantees HTTP/2 works, but does it guarantee HTTP/1 does *not* work?

Reply:
How does it work when I have `HTTPS` and `H2` (or `HTTP` and `H2C`) on the same port?

If we say its a conflict, then that means that if I want to do h1 and h2 on the same port, I need to name the port `HTTP`, not `H2C`. So we will need to tell users "If you want to use HTTP2 over clear text, do NOT set protocol: h2c".

If its not a conflict and they can merge, then that would imply we can accept HTTP/1.1 over the H2C/HTTP2 protocols which seems unexpected?

Reply:
> Doesn't the spec not allow that?

The word "or" is doing a lot of work here. If we should be interpreting this to mean "Accepts HTTP/1.1 or (HTTP/1.1. and HTTP/2)", then I suppose so.

> How does it work when I have HTTPS and H2 (or HTTP and H2C) on the same port?

I would expect `HTTPS` and `H2` on the same port to mean that the port would accept any of the following over TLS:

- HTTP/1.1
- HTTP/2 via upgrade from HTTP/1.1
- HTTP/2 directly (i.e. without upgrade from HTTP/1.1)

Whereas `H2` on its own would guarantee only:

- HTTP/2 directly (i.e. without upgrade from HTTP/1.1)

While `HTTPS` would guarantee:

- HTTP/1.1
- HTTP/2 via upgrade from HTTP/1.1

> If its not a conflict and they can merge, then that would imply we can accept HTTP/1.1 over the H2C/HTTP2 protocols which seems unexpected?

I don't think this is true. A listener with both `HTTPS` and `H2` that accepts HTTP/1.1 would be doing so under the capabilities of `HTTPS`, not of `H2`. This seems similar to what you'd expect for a port that accepts both TCP and HTTP (or is this not allowed?)


More generally, I need to solve two problems here:

- Providing an indication that an implementation definitely supports HTTP/2 (this may be solved already, as you pointed out, depending on how you read the current spec for HTTPS)
- Providing an indication that an implementation definitely accepts HTTP/2 without first upgrading from HTTP/1. Not all servers do this.

If you can propose an alternative that accomplishes these two goals, I'd be more than happy to adopt it.

Reply:
When you say upgrade are you referring to https://en.wikipedia.org/wiki/HTTP/1.1_Upgrade_header#Use_with_HTTP/2 or something else? Since "The HTTP Upgrade mechanism is used only for cleartext HTTP2 (h2c)"

Reply:
Given both the HTTP and HTTPS protocol, in the current spec, allow HTTP2, which don't we just say that GRPC can only bind to HTTP/HTTPS, and the GRPCRoutes only accept HTTP/2 requests.

I don't see why we need 'Providing an indication that an implementation definitely supports HTTP/2'? if the implementation accepts GRPCRoutes then they must accept HTTP/2 (we can declare that in the spec if needed)

Reply:
> When you say upgrade are you referring to https://en.wikipedia.org/wiki/HTTP/1.1_Upgrade_header#Use_with_HTTP/2

Yes.

> if the implementation accepts GRPCRoutes then they must accept HTTP/2 (we can declare that in the spec if needed)

I was under the impression that this sort of requirement was expected to be expressed through the `ProtocolType`. I have no problem with this just being part of the spec for GRPCRoute.

I have switched to making this requirement in the spec for GRPCRoute and have removed the `H2` and `H2C` ProtocolTypes entirely.